tests/auth_tests/test_hashers.py

```diff
@@ -359,3 +364,60 @@ def test_load_library_importerror(self):
         with six.assertRaisesRegex(self, ValueError,
                 "Couldn't load 'PlainHasher' algorithm library: No module named '?plain'?"):
             PlainHasher()._load_library()
+
+
+@skipUnless(argon2, "argon2-cffi not installed")
+@override_settings(PASSWORD_HASHERS=PASSWORD_HASHERS)
+class TestUtilsHashPassArgon2(SimpleTestCase):
+
+    def test_argon2(self):
+        encoded = make_password('lètmein', hasher='argon2')
+        self.assertTrue(is_password_usable(encoded))
+        self.assertTrue(encoded.startswith('argon2$'))
+        self.assertTrue(check_password('lètmein', encoded))
+        self.assertFalse(check_password('lètmeinz', encoded))
+        self.assertEqual(identify_hasher(encoded).algorithm, "argon2")
+        # Blank passwords
+        blank_encoded = make_password('', hasher='argon2')
+        self.assertTrue(blank_encoded.startswith('argon2$'))
+        self.assertTrue(is_password_usable(blank_encoded))
+        self.assertTrue(check_password('', blank_encoded))
+        self.assertFalse(check_password(' ', blank_encoded))
+
+    def test_argon2_upgrade(self):
+        hasher = get_hasher('argon2')
+        self.assertEqual('argon2', hasher.algorithm)
+
+        for attr, summary_key, new_value in (
+                ('time_cost', 'time cost', 1),
+                ('memory_cost', 'memory cost', 16),
+                ('parallelism', 'parallelism', 1)):
+            self.assertNotEqual(getattr(hasher, attr), new_value)
+
+            old_value = getattr(hasher, attr)
+            try:
+                # Generate hash with attr set to 1
+                setattr(hasher, attr, new_value)
+                encoded = make_password('letmein', hasher='argon2')
+                attr_value = hasher.safe_summary(encoded)[summary_key]
+                self.assertEqual(attr_value, new_value)
+
+                state = {'upgraded': False}
+
+                def setter(password):
+                    state['upgraded'] = True
+
+                # Check that no upgrade is triggered.
+                self.assertTrue(check_password('letmein', encoded, setter,
```

Comment:
1 line is okay --- we prefer longer lines up to 119 characters when it helps readability.